tests/components/yale/test_init.py

```diff
@@ -0,0 +1,238 @@
+    STATE_LOCKED,
+    STATE_ON,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.setup import async_setup_component
+
+from .mocks import (
+    _create_yale_with_devices,
+    _mock_doorsense_enabled_yale_lock_detail,
+    _mock_doorsense_missing_yale_lock_detail,
+    _mock_inoperative_yale_lock_detail,
+    _mock_lock_with_offline_key,
+    _mock_operative_yale_lock_detail,
+)
+
+from tests.typing import WebSocketGenerator
+
+
+async def test_yale_api_is_failing(hass: HomeAssistant) -> None:
+    """Config entry state is SETUP_RETRY when yale api is failing."""
+
+    config_entry, socketio = await _create_yale_with_devices(
+        hass,
+        authenticate_side_effect=YaleApiError(
+            "offline", ClientResponseError(None, None, status=500)
+        ),
+    )
+    assert config_entry.state is ConfigEntryState.SETUP_RETRY
+
+
+async def test_yale_is_offline(hass: HomeAssistant) -> None:
+    """Config entry state is SETUP_RETRY when yale is offline."""
+
+    config_entry, socketio = await _create_yale_with_devices(
+        hass, authenticate_side_effect=TimeoutError
+    )
+
+    assert config_entry.state is ConfigEntryState.SETUP_RETRY
+
+
+async def test_yale_late_auth_failure(hass: HomeAssistant) -> None:
+    """Test we can detect a late auth failure."""
+    config_entry, socketio = await _create_yale_with_devices(
+        hass,
+        authenticate_side_effect=InvalidAuth(
+            "authfailed", ClientResponseError(None, None, status=401)
+        ),
+    )
+
+    assert config_entry.state is ConfigEntryState.SETUP_ERROR
+    flows = hass.config_entries.flow.async_progress()
+
+    assert flows[0]["step_id"] == "pick_implementation"
+
+
+async def test_unlock_throws_yale_api_http_error(hass: HomeAssistant) -> None:
+    """Test unlock throws correct error on http error."""
+    mocked_lock_detail = await _mock_operative_yale_lock_detail(hass)
+    aiohttp_client_response_exception = ClientResponseError(None, None, status=400)
+
+    def _unlock_return_activities_side_effect(access_token, device_id):
+        raise YaleApiError(
+            "This should bubble up as its user consumable",
+            aiohttp_client_response_exception,
+        )
+
+    await _create_yale_with_devices(
+        hass,
+        [mocked_lock_detail],
+        api_call_side_effects={
+            "unlock_return_activities": _unlock_return_activities_side_effect
+        },
+    )
+    last_err = None
+    data = {ATTR_ENTITY_ID: "lock.a6697750d607098bae8d6baa11ef8063_name"}
+    try:
+        await hass.services.async_call(LOCK_DOMAIN, SERVICE_UNLOCK, data, blocking=True)
+    except HomeAssistantError as err:
+        last_err = err
+    assert str(last_err) == (
+        "A6697750D607098BAE8D6BAA11EF8063 Name: This should bubble up as its user"
+        " consumable"
+    )
+
```

Comment:
Should we use pytest.raises?

Reply:
- [x] yale - 108398e8fdc16bad93429bf9c81ba95213ff4cb8
- [ ] august -